EasyRPG/Player
# Implement VFS and ZIP reading support (FS [3/4])

Fix #1085 
Fix #2491

TODO/FIXME:

- [x] Platform Test: Emscripten
- [x] Platform Test: Linux
- [x] Tree has lots of duplications in the cache (path with \\, with /, with trailing /)
- [x] Improve the ZIP FS, the code is very old, still from ChrisBreiti from ~5 years ago
- [x] Creating Sub-Filesystems in the GameBrowser leaks memory (FS never deleted)
- [x] Unit tests fail with autotools (but work with CMake o_O)
- [x] Refactor RTP
- [x] Crash when the game is directly in the root of a ZIP archive
- [x] ZIP parser does not notice intermediate folders
- ~~Restarting lacks some cleanup (e.g. unload core in RetroArch and loading again segfaults)~~ Cannot reproduce. RetroArch bug?

--------

Major change:
Before accessing a file you must decide if it is a game file (``FileFinder::Game()``), a save file (``FileFinder::Save()``) or something outside (``FileFinder::Root()``). These return a subview into the corresponding folders. Adding something like ``FileFinder::Config()`` to get the global config dir would make sense here.

--------

Loading games from ZIP files:
The Filesystem ```Create`` function will attempt to access the path directly and when this fails will split it in subcomponents and attempt to construct a VFS path.
Chaining is supported, this means a ZIP can contain a ZIP (though this becomes slow when you chain too much).

To boot a game from a zip simply specify the path as usual:

```
--project-path some_path/games.zip/TestGame-2000
```

---------

Saving inside zip files

ZIPs are obviously not writable (at least not in a sane way).
When the FS does not support the ``Write`` Feature (and ``--save-path`` is not set) then the FileFinder will construct a savepath next to the non-writable file and create the folders, e.g.:

- Game path is ``some_path/games.zip/TestGame-2000``
- Save path will be ``some_path/games.zip.save/TestGame-2000``

- Game path is ``some_path/tons_of_games.zip/TestGames.zip/TestGame-2000``
- Save path will be ``some_path/tons_of_games.zip.save/TestGames.zip/TestGame-2000``

---------

New classes:

- ``Filesystem`` - Base class of all filesystems`
- ``FilesystemView`` - Holds a reference to a filesystem and forwards all calls to it (this pattern makes storing references to file systems simpler because the child class does not matter)
- ``NativeFilesystem`` - Forwards all calls to the API of the operating system (used for the Root filesystem)
- ``ZipFilesystem - Provides the Filesystem Interface for ZIP files. The encoding is auto-detected.

-----------

I decided again to keep ``FileFinder::FindDefault`` and friends, at least until the gigantic PRs such as battle are in. Want to prevent rebase chaos.

-----------

As you can see I use ``DirectoryTree`` as part of the filesystem but the ``DirectoryTreeView`` I just invented is already deleted again. I used the knowledge gained here to invent ``FilesystemView`` and search replaced it everywhere. (The API is actually almost the same as ``DirectoryTreeView`` but it is layered over a VFS)
